### PR TITLE
*: fix the issue of executing DDL after executing SQL failure in txn

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve? 
Fix the issue of the following case:
1. Execute the SQL of "begin";
2. A SQL that will fail to execute;
3. Execute DDL.

Specific examples are as follows:
```
create table t (i int key);
insert into t values (1);
begin;
insert into t values (1);
truncate table t;
```
The goroutine stack as follows:
```
2018/10/24 17:24:22.168 conn.go:427: [error] lastCmd truncate table t, runtime error: invalid memory address or nil pointer dereference, goroutine 388 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0xc426fe2410, 0xc427091dff)
	/Users/xia/workspace/src/github.com/pingcap/tidb/server/conn.go:425 +0x10e
panic(0x1db59c0, 0x2a95190)
	/Users/xia/go/src/runtime/panic.go:505 +0x229
github.com/pingcap/tidb/executor.(*DDLExec).toErr(0xc4279be640, 0x2099240, 0xc42783c840, 0xc42783c840, 0x1b96a03)
	/Users/xia/workspace/src/github.com/pingcap/tidb/executor/ddl.go:58 +0x175
github.com/pingcap/tidb/executor.(*DDLExec).Next(0xc4279be640, 0x31bc488, 0xc4279d8680, 0xc427a79350, 0x2099240, 0xc42783c840)
	/Users/xia/workspace/src/github.com/pingcap/tidb/executor/ddl.go:94 +0x267
github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor(0xc427f00f00, 0x31bc488, 0xc4279d8680, 0x20d0e20, 0xc426f147e0, 0x20b1fa0, 0xc4279be640, 0x0, 0x0, 0x0, ...)
	/Users/xia/workspace/src/github.com/pingcap/tidb/executor/adapter.go:272 +0x157
github.com/pingcap/tidb/executor.(*ExecStmt).Exec(0xc427f00f00, 0x31bc488, 0xc4279d8680, 0x0, 0x0, 0x0, 0x0)
...
```

### What is changed and how it works?
The `NewTxn` will commit the original transaction, then begin a new transaction.
The reason is that we used the transaction to do something after we failed to execute `NewTxn`. So we can move `NewTxn` before using the transaction.

### Check List 

Tests 

 - Unit test

Code changes
Side effects
Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @jackysp @winkyao  @tiancaiamao

File: executor/ddl_test.go [hunk 1]
```diff
@@ -46,6 +46,23 @@ func (s *testSuite) TestTruncateTable(c *C) {
...
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("create table t (i int key);")
+	tk.MustExec("insert into t values (1);")
+	tk.MustExec("begin;")
+	tk.MustExec("insert into t values (1);")
+	_, err := tk.Exec("truncate table t;")
+	c.Assert(err.Error(), Equals, "[kv:1062]Duplicate entry '1' for key 'PRIMARY'")
+	result := tk.MustQuery("select count(*) from t")
+	result.Check(testkit.Rows("1"))
```
Comment (jackysp) [inline executor/ddl_test.go]: Both the previous transaction and the DDL failed. Is it expected?

File: executor/ddl.go [hunk 2]
```diff
@@ -67,6 +67,11 @@ func (e *DDLExec) Next(ctx context.Context, chk *chunk.Chunk) (err error) {
 		return nil
 	}
 	e.done = true
+
+	// For every DDL, we must commit current transaction.
+	if err := e.ctx.NewTxn(); err != nil {
```
Comment (ngaut) [inline executor/ddl.go]: I am confused, according to the comments, i am expecting to call something like "Commit".

File: executor/ddl_test.go (hunk 1 above)
Comment (zimulala) [reply executor/ddl_test.go]: Yes.

**Comment by zimulala** [comment]
@jackysp 
>  commit could fail in TiDB, which is nearly not happened in MySQL.
This PR will fix the bug. Regarding this compatibility issue, the original TiDB is handled differently than MySQL. So we can discuss it later.

**Pull request merged**